sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -290,16 +301,20 @@ case class GreaterThan(left: Expression, right: Expression) extends BinaryCompar
 case class GreaterThanOrEqual(left: Expression, right: Expression) extends BinaryComparison {
   def symbol = ">="
 
-  lazy val ordering = {
-    if (left.dataType != right.dataType) {
-      throw new TreeNodeException(this,
-        s"Types do not match ${left.dataType} != ${right.dataType}")
-    }
-    left.dataType match {
-      case i: NativeType => i.ordering.asInstanceOf[Ordering[Any]]
-      case other => sys.error(s"Type $other does not support ordered operations")
+  override lazy val resolved =
+    left.resolved && right.resolved &&
+    left.dataType == right.dataType &&
+    (left.dataType.isInstanceOf[NativeType] || left.dataType.isInstanceOf[NullType])
+
+  val ordering =
+    if (resolved) {
+      left.dataType match {
+        case n: NativeType => n.ordering.asInstanceOf[Ordering[Any]]
+        case n: NullType => UnresolvedOrdering
+      }
+    } else {
+      UnresolvedOrdering
     }
```

Comment:
Perhaps we could add helper methods to Expression for checking if things are numeric and for finding the correct ordering / numeric types.  This pattern seems to be repeated quite a bit.